Fixups, patsy consistency, and support for predict_proba

ceholden opened the pull request:
Thanks for this work -- it's made my life a lot easier when running scikit-learn models on data, often derivations or subsets of data in `xarray.Dataset` objects.

This PR addresses four types of changes I've made to better use `patsylearn` in my work:

1. Improve `setup.py` by fixing the package name (was `dask`) and by adding `install_requires` and `extras_require` for `pandas` support. Support for `pandas` is now optional for parity with `patsy`
2. Change the `return_type` specification to be consistent with `patsy` language. Prior to PR, the return for `patsy.DesignMatrix` was given as `ndarray`, but the `patsy` convention is to use `matrix`. The checking for the return type in this library is specific to the `dataframe` return type, so code that uses the `matrix` specification will slip through and return the expected `patsy.DesignMatrix` and be backward-compatible.
3. Add wrappers for `predict_proba` and `predict_log_proba` to the `PatsyModel` class.
4. When the `return_type` is `dataframe`, return the output of the `sklearn` estimator as a `DataFrame`. Previously, when asked to return a `DataFrame` via `return_type="dataframe"`, all methods would return a `np.ndarray` right from `sklearn` and the user would have to re-create this output into a `DataFrame` (see [SO question for example](https://stackoverflow.com/questions/35723472/how-to-use-sklearn-fit-transform-with-pandas-and-return-dataframe-instead-of-num)). This PR changes this behavior by returning a `DataFrame` where the index is from the input `X` and the column names are usually some derivation from the classes estimated.


The first change is just organizational, the second is perhaps pedantic but I think it is good to be consistent, and the third adds new code. The fourth change might change the return type for pre-existing users. I'm not sure what the install base is, but it looks like @sashaostr has incorporated it into their work as well as I have, so I'm especially happy to take feedback on this fourth change and in general.

ceholden committed:
Change package name (dask is taken)

ceholden committed:
Use patsy's 'matrix' return for np.array-like

patsy.design_info.DesignMatrix passes isinstance for
np.ndarray and simplifies code while being more consistent
with known patsy terminology. Can np.array the return values
later if needed

ceholden committed:
predict and predict_proba can now return dataframe

Be consistent in returning pd.DataFrame from these methods
when return_type is 'dataframe'. DataFrame column names are
taken from the design_y_ and estimator_.classes_ for
predict and predict_proba, respectively.

ceholden committed:
Add predict_log_proba handler

ceholden committed:
Make Pandas optional install

ceholden committed:
Test predict_proba/log_proba

ceholden committed:
Return DF for decision_func/transform when asked

ceholden committed:
Don't use prefix for df column names

ceholden committed:
Bail on column names for transform

Transform doesn't necessarily return same number of features
as input, and it's being removed in 0.19

amueller commented on patsylearn/patsy_adaptor.py:
```diff
@@ -117,8 +124,34 @@ def predict(self, data):
         """
-        X = np.array(dmatrix(self.design_X_, data))
-        return self.estimator_.predict(X)
+        X = dmatrix(self.design_X_, data, return_type=self.return_type)
```
Is that how patsy does it? Changing the return type based on what libraries are installed seems very fragile to me.

amueller commented on patsylearn/patsy_adaptor.py:
```diff
@@ -295,7 +336,4 @@ def transform(self, data):
-            return dmatrix(self.design_, data, return_type='dataframe')
-        else:
-            return np.array(dmatrix(self.design_, data))
+        return dmatrix(self.design_, data, return_type=self.return_type)
```
Will this work with pipelines? The idea was to return a numpy array as that is the default language of scikit-learn.

amueller commented on patsylearn/tests/test_patsy_model.py:
```diff
@@ -113,7 +119,10 @@ def test_stateful_transform():
+
 def test_stateful_transform_dataframe():
+    if not HAS_PANDAS:
+        return
```
raise SkipTest?

amueller reviewed (commented):
Thanks, this looks great. I'm not entirely sure about the return type and the optional conversion to dataframes, though. I'm ok with a lazy import, but as I said inline, changing the return type of a function based on what's installed seems very fragile. I'd rather raise an error. We could change the default return type possibly?

ceholden committed:
Use SkipTest for pandas-related tests if not installed

ceholden committed:
Raise ImportError if request dataframe w/o pandas

ceholden committed:
Remove transform test for RFC and SVC

RFC.transform will be deprecated and SVC.transform
doesn't exist

ceholden committed:
Add (failing) Pipeline test

ceholden committed:
Use column name subset for feature selectors

Feature selecting transformers have `get_support` method

ceholden committed:
Support get/set_params functions by wrapping

ceholden committed:
Fix missing return on PatsyModel.transform

ceholden committed:
Make est.score look like sklearn for pipeline compat

ceholden committed:
Test passing 'y' to Pipeline.fit

ceholden commented:
Thanks for the comments and review.

I've modified the tests to raise `SkipTest` and raised an `ImportError` if one tries to use `return_type='dataframe'` without Pandas installed during the `__init__` function. I agree about wanting to see an exception rather than a change in return type and it is probably clearer to see this exception early on. Without Pandas installed, a user requesting a "dataframe" would eventually reach:

``` python
PatsyError: pandas.DataFrame was requested, but pandas is not installed
```

I began trying to use these wrappers in a `Pipeline` by testing with the `anova_svm` example in the `Pipeline` docstring. There are a few obstacles I fixed (wrapping `set_params`, setting column names correctly for transforms that do feature selection), but I wasn't able to run the pipeline using the `return_type='matrix'` option. Passing around Pandas DataFrames seems to be working through scikit-learn support for datatypes using the `__array__` protocol.

Taking the example of running a `StandardScaler` and then a `SVC` in a `Pipeline`:

When using `matrix` return type, the `StandardScalar.transform` will return "X" as a numpy array that is no longer "dict-like" when it is passed to `SVC.fit`. It's actually the `dataframe` return type that has the best compatibility within scikit-learn Pipelines. I've added this example as a [test here](https://github.com/ceholden/patsylearn/blob/master/patsylearn/tests/test_patsy_model.py#L205)

There are still two rough edges when using Pipelines that I have found:

1. One needs to pass both "data" (e.g., a DataFrame, `data`) and the "y" (e.g., `data['y']`) to the `pipeline.fit` method. This happens because the transform steps will only return a dataframe with the columns related to `X` rather than `y`, causing the `dmatrices` call to fail. Passing `y` will allow this value to be passed through the environment capture mechanism in Patsy.
2. There needs to be some logic added when using feature selection transformations in a Pipeline because subsequent steps will still look toward the entire feature set (e.g., the entire right hand side of the Patsy equation), but the feature selecting transform will be missing some of these features. For example, a `SelectKBest(f_regression, k=2)` running on `y ~ x1 + x2 + x3` would return 2 features in X that aren't determinable prior to setting up `SVC` to run on the same equation.

I'd be happy to work toward smoothing over these edges, especially if you have any advice for how to proceed.

Cheers

amueller commented:
I'm sorry for the delay on looking at this. It's still on my todo list but I'm pretty busy right now, sorry.

jnothman commented on setup.py:
```diff
@@ -4,7 +4,8 @@
 from os.path import exists
 from setuptools import setup
 
-setup(name='dask',
```
not naming the package dask would be a good start :P

amueller replied:
I don't even know why I put that there?!
